src/pcre2_compile.c

```diff
@@ -4322,14 +4315,24 @@ while (ptr < ptrend)
                 goto FAILED;
                 }
 
-              if (!read_name(&ptr, ptrend, utf, terminator, &offset, &name,
-                  &namelen, &errorcode, cb)) goto FAILED;
+              if (!read_name(&ptr, ptrend, utf, terminator, &next_offset,
+                  &name, &namelen, &errorcode, cb)) goto FAILED;
 
-              *parsed_pattern++ = meta ? META_SCS_NEXT_NAME : META_SCS_NAME;
-              *parsed_pattern++ = namelen;
-              PUTOFFSET(offset, parsed_pattern);
+              meta = META_SCS_NAME;
+              }
+
+            PCRE2_ASSERT(next_offset > 0);
+            if (offset == 0 || (next_offset - offset) >= 0x10000)
+              {
+              *parsed_pattern++ = META_OFFSET;
+              PUTOFFSET(next_offset, parsed_pattern);
+              offset = next_offset;
               }
 
+            *parsed_pattern++ = meta | (uint32_t)(next_offset - offset);
+            *parsed_pattern++ = namelen;
```

Comment:
It might help a future maintainer to just a tiny comment here:

"We need to use a relative offset which fits in the bottom 16 bits (meta_arg), since we cannot output more that 2x uint32_t values; a numbered group reference could be as short as two characters ("1," including the comma or closing parenthesis), and the META codes must be no longer than the input pattern."